File: src/mobile/android/app/src/main/kotlin/com/example/saasportal/ui/services/ServiceListViewModel.kt
package com.example.saasportal.ui.services

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.saasportal.domain.model.ServiceLaunchResult
import com.example.saasportal.domain.model.Subscription
import com.example.saasportal.domain.repository.PortalRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ServiceListViewModel @Inject constructor(
    private val portalRepository: PortalRepository,
) : ViewModel() {

    private val _uiState = MutableStateFlow<ServiceListUiState>(ServiceListUiState.Loading)
    val uiState: StateFlow<ServiceListUiState> = _uiState.asStateFlow()

    private val _launchResult = MutableStateFlow<ServiceLaunchResult?>(null)
    val launchResult: StateFlow<ServiceLaunchResult?> = _launchResult.asStateFlow()

    init {
        loadServices()
    }

    fun loadServices() {
        viewModelScope.launch {
            _uiState.value = ServiceListUiState.Loading
            try {
                val subscriptions = portalRepository.getSubscriptions()
                _uiState.value = ServiceListUiState.Success(subscriptions)
            } catch (e: Exception) {
                _uiState.value = ServiceListUiState.Error(
                    e.message ?: "サービス一覧の取得に失敗しました",
                )
            }
        }
    }

    fun launchService(serviceCode: String) {
        viewModelScope.launch {
            try {
                val result = portalRepository.launchService(serviceCode)
                _launchResult.value = result
            } catch (e: Exception) {
                // Launch failure is non-critical; keep list visible
            }
        }
    }

    fun clearLaunchResult() {
        _launchResult.value = null
    }
}

sealed interface ServiceListUiState {
    data object Loading : ServiceListUiState
    data class Success(val subscriptions: List<Subscription>) : ServiceListUiState
    data class Error(val message: String) : ServiceListUiState
}

File: src/mobile/ios/SaaSPortal/ViewModels/ServiceListViewModel.swift
import Foundation
import os

private let logger = Logger(subsystem: "com.example.SaaSPortal", category: "ServiceListViewModel")

/// サービス一覧画面の UI 状態。
enum ServiceListUiState {
    case loading
    case success([Subscription])
    case error(String)
}

/// サービス一覧画面の状態管理 ViewModel。
@Observable
final class ServiceListViewModel {
    private(set) var uiState: ServiceListUiState = .loading
    private(set) var launchResult: ServiceLaunchResult?
    var showLaunchAlert = false
    private(set) var isLaunching = false
    private let portalService: PortalServiceProtocol

    init(portalService: PortalServiceProtocol = PortalService()) {
        self.portalService = portalService
    }

    /// 契約サービス一覧を取得する。
    @MainActor
    func loadServices() async {
        uiState = .loading
        do {
            let response = try await portalService.getSubscriptions()
            uiState = .success(response.subscriptions)
        } catch {
            logger.error("❌ loadServices failed: \(error.localizedDescription, privacy: .public)")
            uiState = .error(errorMessage(from: error))
        }
    }

    /// サービスを起動し、結果をアラートで表示する。
    @MainActor
    func launchService(serviceCode: String) async {
        isLaunching = true
        do {
            let result = try await portalService.launchService(serviceCode: serviceCode)
            launchResult = result
            showLaunchAlert = true
        } catch {
            logger.error("❌ launchService(\(serviceCode, privacy: .public)) failed: \(error.localizedDescription, privacy: .public)")
            launchResult = nil
            showLaunchAlert = true
        }
        isLaunching = false
    }

    /// アラートを閉じてランチ結果をクリアする。
    func clearLaunchResult() {
        launchResult = nil
        showLaunchAlert = false
    }

    // MARK: - Private

    private func errorMessage(from error: Error) -> String {
        if let apiError = error as? APIError {
            return apiError.localizedDescription
        }
        return error.localizedDescription
    }
}
